src/duckling/game/drawable/shape-drawable.component.ts

```diff
@@ -0,0 +1,93 @@
+import {
+    Component,
+    Input,
+    Output,
+    EventEmitter,
+    AfterViewInit
+} from '@angular/core';
+import {NgSwitch, NgSwitchCase, NgSwitchDefault} from '@angular/common';
+import {MD_INPUT_DIRECTIVES} from '@angular2-material/input';
+
+import {EnumChoiceComponent, SelectOption, FormLabel} from '../../controls';
+import {immutableAssign} from '../../util/model';
+import {NumberInput} from '../../controls/number-input.component';
+
+import {ShapeDrawable} from './shape-drawable';
+import {GenericShapeComponent} from './generic-shape.component';
+import {Shape, ShapeType} from './shape';
+import {defaultCircle} from './circle';
+import {CircleComponent} from './circle.component';
+import {defaultRectangle} from './rectangle';
+import {RectangleComponent} from './rectangle.component';
+
+@Component({
+    selector: "dk-shape-drawable-component",
+    directives: [
+        GenericShapeComponent,
+        CircleComponent,
+        RectangleComponent,
+        EnumChoiceComponent,
+        NgSwitch,
+        NgSwitchCase,
+        NgSwitchDefault
+    ],
+    template: `
+        <dk-generic-shape-component
+            *ngIf="shapeDrawable.shape?.type !== null"
+            [shape]="shapeDrawable.shape"
+            (shapeChanged)="specificShapeChanged($event)">
+        </dk-generic-shape-component>
+
+        <div [ngSwitch]="shapeDrawable.shape?.type">
+            <dk-enum-choice
+                *ngSwitchDefault
+                [enum]="ShapeType"
+                [selected]="ShapeType.Circle"
+                (addClicked)="onShapeTypePicked($event)">
+            </dk-enum-choice>
+
+            <dk-circle-component
+                *ngSwitchCase="ShapeType.Circle"
+                [circle]="shapeDrawable.shape"
+                (circleChanged)="onShapeChanged($event)">
+            </dk-circle-component>
+
+            <dk-rectangle-component
+                *ngSwitchCase="ShapeType.Rectangle"
+                [rectangle]="shapeDrawable.shape"
+                (rectangleChanged)="onShapeChanged($event)">
+            </dk-rectangle-component>
+        </div>
+    `
+})
+export class ShapeDrawableComponent {
+    // hoist ShapeType so template can access it
+    ShapeType = ShapeType;
+
+    @Input() shapeDrawable : ShapeDrawable;
+    @Output() drawableChanged = new EventEmitter<ShapeDrawable>();
+
+    specificShapeChanged(newShape : Shape) {
+        this.drawableChanged.emit(immutableAssign(this.shapeDrawable, {shape: newShape}));
+    }
+
+    onShapeTypePicked(pickedType : ShapeType) {
+        var patch : Shape = this.getDefaultShape(pickedType);
+        this.drawableChanged.emit(immutableAssign(this.shapeDrawable, {shape: patch}));
+    }
+
+    onShapeChanged(shape : Shape) {
+        this.drawableChanged.emit(immutableAssign(this.shapeDrawable, {shape: shape}));
+    }
+
+    private getDefaultShape(type : ShapeType) : Shape {
```

Comment:
I don't like switching on these (also done in getDefaultDrawable) but I can't use the default attribute service because that's specialized to attributes. Maybe we could change that service to be based on a more general default service we could use for these scenarios.